### PR TITLE
[rtsan] Re-enable rtsan tests

### PR DESCRIPTION
**Pull request by vitalybuka**
Follow up to #92460

DEPS llvm_gtest is not used by compiler-rt,
compiler-rt compiles them with COMPILER_RT_GOOGLETEST_SOURCES.

This reverts commit e217f98ac3ee55755ee6ea3b97551cf8283528c9.

### CODE REVIEW

**Comment by vitalybuka** [comment]
@cjappl

File: compiler-rt/lib/rtsan/tests/CMakeLists.txt [hunk 1]
```diff
@@ -94,7 +94,6 @@ foreach(arch ${RTSAN_TEST_ARCH})
     COMPILE_DEPS ${RTSAN_UNITTEST_HEADERS}
     SOURCES ${RTSAN_NOINST_TEST_SOURCES}
             ${COMPILER_RT_GOOGLETEST_SOURCES}
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/tests/CMakeLists.txt:96]: Will you remove this dependency in the commented out line around 60 as well? Just want to make sure we don't repeat this mistake when that is commented out

**Review by cjappl — CHANGES_REQUESTED** [review]
We also have the failure on fuchsia and windows as outlined here:

https://github.com/llvm/llvm-project/pull/92460#issuecomment-2218652616

Fuchsia we can "fix" I believe via this patch that disables the problematic interceptors for now:

```
diff --git a/compiler-rt/lib/rtsan/rtsan_interceptors.cpp b/compiler-rt/lib/rtsan/rtsan_interceptors.cpp
index 3a65f9d3f779..ad9a97bdae7c 100644
--- a/compiler-rt/lib/rtsan/rtsan_interceptors.cpp
+++ b/compiler-rt/lib/rtsan/rtsan_interceptors.cpp
@@ -42,6 +42,8 @@ void ExpectNotRealtime(const char *intercepted_function_name) {
 
 // Filesystem
 
+#if !SANITIZER_FUCHSIA
+
 INTERCEPTOR(int, open, const char *path, int oflag, ...) {
   // TODO Establish whether we should intercept here if the flag contains
   // O_NONBLOCK
@@ -70,6 +72,8 @@ INTERCEPTOR(int, openat, int fd, const char *path, int oflag, ...) {
   return result;
 }
 
+#endif //!SANITIZER_FUCHSIA
+
 INTERCEPTOR(int, creat, const char *path, mode_t mode) {
   // TODO Establish whether we should intercept here if the flag contains
   // O_NONBLOCK
@@ -359,8 +363,11 @@ void __rtsan::InitializeInterceptors() {
   INTERCEPT_FUNCTION(pvalloc);
 #endif
 
+#ifndef SANITIZER_FUCHSIA
   INTERCEPT_FUNCTION(open);
   INTERCEPT_FUNCTION(openat);
+#endif //SANITIZER_FUCHSIA
+
   INTERCEPT_FUNCTION(close);
   INTERCEPT_FUNCTION(fopen);
   INTERCEPT_FUNCTION(fread);
diff --git a/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp b/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
index f5b016089087..e250df41b53a 100644
--- a/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
+++ b/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
@@ -172,6 +172,7 @@ TEST(TestRtsanInterceptors, NanosleepDiesWhenRealtime) {
     Filesystem
 */
 
+#if !SANITIZER_FUCHSIA
 TEST_F(RtsanFileTest, OpenDiesWhenRealtime) {
   auto func = [this]() { open(GetTemporaryFilePath(), O_RDONLY); };
   ExpectRealtimeDeath(func, "open");
@@ -197,6 +198,7 @@ TEST_F(RtsanFileTest, OpenCreatesFileWithProperMode) {
   // Mask st_mode to get permission bits only
   ASSERT_THAT(st.st_mode & 0777, Eq(mode));
 }
+#endif // !SANITIZER_FUCHSIA
 
 TEST_F(RtsanFileTest, CreatDiesWhenRealtime) {
   auto func = [this]() { creat(GetTemporaryFilePath(), S_IWOTH | S_IROTH); };
```

Windows we can just disable the entire sanitizer. the guts rely on pthread, so there is no chance of it working at this point, possibly in the future

**Comment by cjappl** [comment]
Sincere thanks for your help @vitalybuka , really appreciate it as an LLVM newbie

**Comment by vitalybuka** [comment]
> We also have the failure on fuchsia and windows as outlined here:
> 
> [#92460 (comment)](https://github.com/llvm/llvm-project/pull/92460#issuecomment-2218652616)
> 
> Fuchsia we can "fix" I believe via this patch that disables the problematic interceptors for now:
> 
> ```
> diff --git a/compiler-rt/lib/rtsan/rtsan_interceptors.cpp b/compiler-rt/lib/rtsan/rtsan_interceptors.cpp
> index 3a65f9d3f779..ad9a97bdae7c 100644
> --- a/compiler-rt/lib/rtsan/rtsan_interceptors.cpp
> +++ b/compiler-rt/lib/rtsan/rtsan_interceptors.cpp
> @@ -42,6 +42,8 @@ void ExpectNotRealtime(const char *intercepted_function_name) {
>  
>  // Filesystem
>  
> +#if !SANITIZER_FUCHSIA
> +
>  INTERCEPTOR(int, open, const char *path, int oflag, ...) {
>    // TODO Establish whether we should intercept here if the flag contains
>    // O_NONBLOCK
> @@ -70,6 +72,8 @@ INTERCEPTOR(int, openat, int fd, const char *path, int oflag, ...) {
>    return result;
>  }
>  
> +#endif //!SANITIZER_FUCHSIA
> +
>  INTERCEPTOR(int, creat, const char *path, mode_t mode) {
>    // TODO Establish whether we should intercept here if the flag contains
>    // O_NONBLOCK
> @@ -359,8 +363,11 @@ void __rtsan::InitializeInterceptors() {
>    INTERCEPT_FUNCTION(pvalloc);
>  #endif
>  
> +#ifndef SANITIZER_FUCHSIA
>    INTERCEPT_FUNCTION(open);
>    INTERCEPT_FUNCTION(openat);
> +#endif //SANITIZER_FUCHSIA
> +
>    INTERCEPT_FUNCTION(close);
>    INTERCEPT_FUNCTION(fopen);
>    INTERCEPT_FUNCTION(fread);
> diff --git a/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp b/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
> index f5b016089087..e250df41b53a 100644
> --- a/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
> +++ b/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
> @@ -172,6 +172,7 @@ TEST(TestRtsanInterceptors, NanosleepDiesWhenRealtime) {
>      Filesystem
>  */
>  
> +#if !SANITIZER_FUCHSIA
>  TEST_F(RtsanFileTest, OpenDiesWhenRealtime) {
>    auto func = [this]() { open(GetTemporaryFilePath(), O_RDONLY); };
>    ExpectRealtimeDeath(func, "open");
> @@ -197,6 +198,7 @@ TEST_F(RtsanFileTest, OpenCreatesFileWithProperMode) {
>    // Mask st_mode to get permission bits only
>    ASSERT_THAT(st.st_mode & 0777, Eq(mode));
>  }
> +#endif // !SANITIZER_FUCHSIA
>  
>  TEST_F(RtsanFileTest, CreatDiesWhenRealtime) {
>    auto func = [this]() { creat(GetTemporaryFilePath(), S_IWOTH | S_IROTH); };
> ```
> 
> Windows we can just disable the entire sanitizer. the guts rely on pthread, so there is no chance of it working at this point, possibly in the future

Oh, as I remember fuchia does not have interceptors at all

**Comment by cjappl** [comment]
I think adding this patch would disable on windows and fuchsia, and we can evaluate in the future to enable them:


```
diff --git a/compiler-rt/cmake/config-ix.cmake b/compiler-rt/cmake/config-ix.cmake
index 1130f9f721df..6df15c6cec17 100644
--- a/compiler-rt/cmake/config-ix.cmake
+++ b/compiler-rt/cmake/config-ix.cmake
@@ -751,7 +751,8 @@ else()
   set(COMPILER_RT_HAS_ASAN FALSE)
 endif()
 
-if (COMPILER_RT_HAS_SANITIZER_COMMON AND RTSAN_SUPPORTED_ARCH)
+# RTSan does not run on windows or fuschia for now
+if (COMPILER_RT_HAS_SANITIZER_COMMON AND RTSAN_SUPPORTED_ARCH AND NOT OS_NAME MATCHES "Windows|Fuchsia")
   set(COMPILER_RT_HAS_RTSAN TRUE)
 else()
   set(COMPILER_RT_HAS_RTSAN FALSE)
diff --git a/compiler-rt/lib/rtsan/CMakeLists.txt b/compiler-rt/lib/rtsan/CMakeLists.txt
index bd7358e86e59..3cc27a2a8338 100644
--- a/compiler-rt/lib/rtsan/CMakeLists.txt
+++ b/compiler-rt/lib/rtsan/CMakeLists.txt
@@ -25,66 +25,68 @@ set(RTSAN_LINK_LIBS
   ${COMPILER_RT_UNWINDER_LINK_LIBS}
   ${COMPILER_RT_CXX_LINK_LIBS})
 
-if(APPLE)
-  add_compiler_rt_object_libraries(RTRtsan
-    OS ${SANITIZER_COMMON_SUPPORTED_OS}
-    ARCHS ${RTSAN_SUPPORTED_ARCH}
-    SOURCES ${RTSAN_CXX_SOURCES}
-    ADDITIONAL_HEADERS ${RTSAN_HEADERS}
-    CFLAGS ${RTSAN_CFLAGS}
-    DEPS ${RTSAN_DEPS})
-else()
-  add_compiler_rt_object_libraries(RTRtsan
-    ARCHS ${RTSAN_SUPPORTED_ARCH}
-    SOURCES ${RTSAN_CXX_SOURCES}
-    ADDITIONAL_HEADERS ${RTSAN_HEADERS}
-    CFLAGS ${RTSAN_CFLAGS}
-    DEPS ${RTSAN_DEPS})
-  add_compiler_rt_object_libraries(RTRtsan_preinit
-    ARCHS ${RTSAN_SUPPORTED_ARCH}
-    SOURCES ${RTSAN_PREINIT_SOURCES}
-    ADDITIONAL_HEADERS ${RTSAN_HEADERS}
-    CFLAGS ${RTSAN_CFLAGS})
-endif()
+if (COMPILER_RT_HAS_RTSAN)
+  if(APPLE)
+    add_compiler_rt_object_libraries(RTRtsan
+      OS ${SANITIZER_COMMON_SUPPORTED_OS}
+      ARCHS ${RTSAN_SUPPORTED_ARCH}
+      SOURCES ${RTSAN_CXX_SOURCES}
+      ADDITIONAL_HEADERS ${RTSAN_HEADERS}
+      CFLAGS ${RTSAN_CFLAGS}
+      DEPS ${RTSAN_DEPS})
+  else()
+    add_compiler_rt_object_libraries(RTRtsan
+      ARCHS ${RTSAN_SUPPORTED_ARCH}
+      SOURCES ${RTSAN_CXX_SOURCES}
+      ADDITIONAL_HEADERS ${RTSAN_HEADERS}
+      CFLAGS ${RTSAN_CFLAGS}
+      DEPS ${RTSAN_DEPS})
+    add_compiler_rt_object_libraries(RTRtsan_preinit
+      ARCHS ${RTSAN_SUPPORTED_ARCH}
+      SOURCES ${RTSAN_PREINIT_SOURCES}
+      ADDITIONAL_HEADERS ${RTSAN_HEADERS}
+      CFLAGS ${RTSAN_CFLAGS})
+  endif()
 
-set(RTSAN_COMMON_RUNTIME_OBJECT_LIBS
-  RTInterception
-  RTSanitizerCommon
-  RTSanitizerCommonLibc
-  RTSanitizerCommonCoverage
-  RTSanitizerCommonSymbolizer)
+  set(RTSAN_COMMON_RUNTIME_OBJECT_LIBS
+    RTInterception
+    RTSanitizerCommon
+    RTSanitizerCommonLibc
+    RTSanitizerCommonCoverage
+    RTSanitizerCommonSymbolizer)
 
-append_list_if(COMPILER_RT_HAS_LIBDL dl RTSAN_LINK_LIBS)
-append_list_if(COMPILER_RT_HAS_LIBRT rt RTSAN_LINK_LIBS)
-append_list_if(COMPILER_RT_HAS_LIBM m RTSAN_LINK_LIBS)
-append_list_if(COMPILER_RT_HAS_LIBPTHREAD pthread RTSAN_LINK_LIBS)
-append_list_if(COMPILER_RT_HAS_LIBLOG log RTSAN_LINK_LIBS)
+  append_list_if(COMPILER_RT_HAS_LIBDL dl RTSAN_LINK_LIBS)
+  append_list_if(COMPILER_RT_HAS_LIBRT rt RTSAN_LINK_LIBS)
+  append_list_if(COMPILER_RT_HAS_LIBM m RTSAN_LINK_LIBS)
+  append_list_if(COMPILER_RT_HAS_LIBPTHREAD pthread RTSAN_LINK_LIBS)
+  append_list_if(COMPILER_RT_HAS_LIBLOG log RTSAN_LINK_LIBS)
 
-add_compiler_rt_component(rtsan)
+  add_compiler_rt_component(rtsan)
 
-if (APPLE)
-  add_weak_symbols("sanitizer_common" WEAK_SYMBOL_LINK_FLAGS)
-  set(RTSAN_LINK_FLAGS ${RTSAN_LINK_FLAGS} ${WEAK_SYMBOL_LINK_FLAGS})
+  if (APPLE)
+    add_weak_symbols("sanitizer_common" WEAK_SYMBOL_LINK_FLAGS)
+    set(RTSAN_LINK_FLAGS ${RTSAN_LINK_FLAGS} ${WEAK_SYMBOL_LINK_FLAGS})
 
-  add_compiler_rt_runtime(clang_rt.rtsan
-    SHARED
-    OS ${SANITIZER_COMMON_SUPPORTED_OS}
-    ARCHS ${RTSAN_SUPPORTED_ARCH}
-    OBJECT_LIBS RTRtsan
-                ${RTSAN_COMMON_RUNTIME_OBJECT_LIBS}
-    LINK_FLAGS ${RTSAN_LINK_FLAGS}
-    LINK_LIBS ${RTSAN_LINK_LIBS}
-    PARENT_TARGET rtsan)
-else()
-  add_compiler_rt_runtime(clang_rt.rtsan
-    STATIC
-    ARCHS ${RTSAN_SUPPORTED_ARCH}
-    OBJECT_LIBS RTRtsan_preinit
-                RTRtsan
-                ${RTSAN_COMMON_RUNTIME_OBJECT_LIBS}
-    LINK_FLAGS ${RTSAN_LINK_FLAGS}
-    CFLAGS ${RTSAN_CFLAGS}
-    PARENT_TARGET rtsan)
+    add_compiler_rt_runtime(clang_rt.rtsan
+      SHARED
+      OS ${SANITIZER_COMMON_SUPPORTED_OS}
+      ARCHS ${RTSAN_SUPPORTED_ARCH}
+      OBJECT_LIBS RTRtsan
+                  ${RTSAN_COMMON_RUNTIME_OBJECT_LIBS}
+      LINK_FLAGS ${RTSAN_LINK_FLAGS}
+      LINK_LIBS ${RTSAN_LINK_LIBS}
+      PARENT_TARGET rtsan)
+  else()
+    add_compiler_rt_runtime(clang_rt.rtsan
+      STATIC
+      ARCHS ${RTSAN_SUPPORTED_ARCH}
+      OBJECT_LIBS RTRtsan_preinit
+                  RTRtsan
+                  ${RTSAN_COMMON_RUNTIME_OBJECT_LIBS}
+      LINK_FLAGS ${RTSAN_LINK_FLAGS}
+      CFLAGS ${RTSAN_CFLAGS}
+      PARENT_TARGET rtsan)
+  endif()
 endif()
 
 if(COMPILER_RT_INCLUDE_TESTS)


```

**Comment by cjappl** [comment]
(to be clear, I don't have a windows or fuchsia machine to test this change on, please be as skeptical as possible - this is just based on the code I see around it, and trying out disabling Darwin)

File: compiler-rt/cmake/config-ix.cmake [hunk 2]
```diff
@@ -751,7 +751,8 @@ else()
   set(COMPILER_RT_HAS_ASAN FALSE)
 endif()
 
-if (COMPILER_RT_HAS_SANITIZER_COMMON AND RTSAN_SUPPORTED_ARCH)
+if (COMPILER_RT_HAS_SANITIZER_COMMON AND RTSAN_SUPPORTED_ARCH AND
+    OS_NAME MATCHES "Android|Darwin|Linux")
```
Comment (vitalybuka) [inline compiler-rt/cmake/config-ix.cmake:755]: @cjappl how about this and let motivated folks with access to those OSes to test and enable when needed.

Comment (cjappl) [reply compiler-rt/cmake/config-ix.cmake:755]: Sounds good to me.

I believe we need to add the conditional check for COMPILER_RT_HAS_RTSAN in compiler-rt/lib/rtsan/CMakeLists.txt. We missed that in the first PR.

With that change, this should be good! I'm going to approve, assuming that you add that conditional in there.

Comment (vitalybuka) [reply compiler-rt/cmake/config-ix.cmake:755]: I actually rather limit to whatever you are willing to work in near future.
What is your platforms of interest?

**Review by cjappl — COMMENTED** [review]
See one last comment before submission, we need to wrap some cmake in the improved conditional.

Thanks again for the help. Do let me know if there are any more failures, I will keep an eye on my indbox

File: compiler-rt/cmake/config-ix.cmake (hunk 2 above)
Comment (cjappl) [reply compiler-rt/cmake/config-ix.cmake:755]: Linux and Darwin are the platforms that we have easiest development access to in ubuntu and darwin. I think starting there and building up would be just fine.

**Review by MaskRay — APPROVED** [review]

File: compiler-rt/cmake/config-ix.cmake (hunk 2 above)
Comment (vitalybuka) [reply compiler-rt/cmake/config-ix.cmake:755]: > I believe we need to add the conditional check for COMPILER_RT_HAS_RTSAN in compiler-rt/lib/rtsan/CMakeLists.txt. We missed that in the first PR.

COMPILER_RT_HAS_${} checked by root lib/ and test/ CMakeLists

**Review by cjappl — APPROVED** [review]
Thanks! Let me know if any more failures come up and I'm happy to help

**Comment by cjappl** [comment]
Looking at this error as we speak

**Comment by vitalybuka** [comment]
-lpthread ?

**Comment by vitalybuka** [comment]
Asan has append_list_if(COMPILER_RT_HAS_LIBPTHREAD -pthread ASAN_UNITTEST_NOINST_LINK_FLAGS) ?

**Comment by vitalybuka** [comment]
As it does not reproduce for me, if you don't mind, I'll leave this one you?
Please send me PR with undo of #98250 and a fix.

**Comment by chapuni** [comment]
I saw https://lab.llvm.org/buildbot/#/builders/41/builds/506 still failing, due to gtest crash.

Let me disable again. (9ae24c9ac94017e15eb827e25c5693418e6cdb4b)

**Comment by cjappl** [comment]
Thanks @chapuni !!

**Pull request merged**